### DISCUSSION TITLE
Option to add custom provider to mlflow gateway

### DISCUSSION THREAD

**Discussion by ktl014** [mlflow/mlflow#10574] [Q&A] [answered]
Hi, I'm interested in trying to add a custom provider to connect an in-house model to mlflow gateway to use for prompt tuning.
Is it possible to add a custom provider when an mlflow gateway instance is executed? How can this be done?

**Comment by B-Step62** [comment, marked as answer]
@ktl014 Thank you for opening this discussion!

One workaround to do that now is to using `mlflow` provider ([doc](https://mlflow.org/docs/latest/llms/gateway/index.html#mlflow-model-serving)). You can log and register your in-house model as [MLflow model](https://mlflow.org/docs/latest/models.html) and pass the serving URL as a backend of MLflow gateway. You can follow [this tutorial](https://github.com/mlflow/mlflow/blob/master/examples/gateway/mlflow_serving/README.md) for setting these up.

That's being said, we acknowledge that this solution is not ideal, for example, if you already have a REST endpoint then serving it as MLflow model is redundant. We don't have concrete plan for supporting this as it will require significant design, but will definitely discuss about this in the next planning.

**Reply by ktl014** [reply]
Awesome, thanks @B-Step62 
Also wanted to follow-up on if there was an easier way to view code-outputs when prompt-engineering (ex: text to sql)?
- It would be great for it to automatically convert the output into code-blocks similar to chatgpt as opposed to just printing out the text directly

**Reply by B-Step62** [reply]
@ktl014 Unfortunately we don't support code rendering on prompt-engineering for now, however, that definitely sounds a neat feature to have:)
Would you mind opening a FR for this? We can pick it up or ask the community to contribute for it.

**Reply by ktl014** [reply]
Done! https://github.com/mlflow/mlflow/issues/10705
